### PR TITLE
[FLINK-15466][table-planner] Fix bug of `FlinkAggregateExpandDistinct…

### PR DESCRIPTION
**Pull request by cshuo**
## What is the purpose of the change

Fix bug of `FlinkAggregateExpandDistinctAggregatesRule`; it'll generate wrong plan for cases where distinct aggregate calls with filter are used.

## Brief change log
  - This PR mainly fix bugs of `FlinkAggregateExpandDistinctAggregatesRule`.

## Verifying this change

This change added tests and can be verified as follows:

  - Added integration tests in `DistinctAggregateITCaseBase` for end-to-end test.
  - Added tests in `DistinctAggregateTest` that validates the plan for cases with distinct aggregate calls with filter.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkAggregateExpandDistinctAggregatesRule.java [hunk 1]
```diff
@@ -471,10 +474,24 @@ private void rewriteUsingGroupingSets(RelOptRuleCall call,
...
+					nodes.add(
 						relBuilder.alias(
-								relBuilder.equals(nodeZ, relBuilder.literal(v)),
-								"$g_" + v));
+							relBuilder.equals(nodeZ, relBuilder.literal(v)),
+							"$g_" + v));
+				} else {
+					RexBuilder rexBuilder = aggregate.getCluster().getRexBuilder();
+					// merge the filter of the distinct aggregate call itself.
+					nodes.add(relBuilder.alias(
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkAggregateExpandDistinctAggregatesRule.java]: it's better to extract `nodes.add(relBuilder.alias(project, "$g_" + v)` as common part

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/agg/DistinctAggregateTest.scala [hunk 2]
```diff
@@ -82,4 +83,39 @@ class DistinctAggregateTest extends TableTestBase {
...
+  }
+
+  @Test
+  def testMultiDistinctAndNonDistinctAggWithFilter(): Unit = {
+    val sqlQuery = "SELECT d, MAX(e), MAX(e) FILTER (WHERE a < 10), COUNT(DISTINCT c),\n" +
+      "COUNT(DISTINCT c) FILTER (WHERE a > 5), COUNT(DISTINCT b) FILTER (WHERE b > 3)\n" +
+      "FROM MyTable2 GROUP BY d"
+    util.verifyPlan(sqlQuery)
+  }
+
```
[the remark this answers is not in the document]
Comment (cshuo) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/agg/DistinctAggregateTest.scala]: > please move these cases to `FlinkAggregateExpandDistinctAggregatesRuleTest`

I noticed there are common distinct tests between `FlinkAggregateExpandDistinctAggregatesRuleTest` and `DistinctAggregateTest`, it's better to reserve them here, and copy them to `FlinkAggregateExpandDistinctAggregatesRuleTest` as well?

[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/agg/DistinctAggregateTest.scala]: actually, It's better to extract a common test case base named `DistinctAggregateTestBase` (like `AggregateReduceGroupingTestBase`)

**Comment by cshuo** [comment]
@godfreyhe I've made some changes. CC @KurtYoung, would you have a look too?

**Pull request merged**